docs(lep): orphaned engine and replica runtime instance cleanup (#10675)
longhorn/longhorn: enhancements/20250331-orphaned-runtime-cleanup.md

# Orphaned Engine And Replica Runtime Instance Cleanup

## Summary

Orphaned runtime instance cleanup identifies unmanaged engine and replica runtime instances on the nodes, and provides a list of the orphan instances on each node. Longhorn by default keeps the instance. Also, it provides a configurable way to clean up such resources automatically.

### Related Issues

[https://github.com/longhorn/longhorn/issues/6764](https://github.com/longhorn/longhorn/issues/6764)

## Motivation

### Goals

- Identify the orphaned engine and replica processes
- The scanning process should not stick to the reconciliation of the controller

> - Provide the user a way to select and trigger the deletion of the orphaned runtime

Reviewer: For wording alignment. Please apply to other related text as well.
```suggestion
- Provide the user a way to select and trigger the deletion of the orphaned runtime instance
```

- Support the global auto-deletion of orphaned runtime

### Non-goals

- Clean up orphaned data stores on the nodes
- Support the per-node auto-deletion of orphaned runtime
- Support the auto-deletion of orphaned runtime exceeding the TTL

## Proposal

> 1. Introduce some new type of existing CRD `orphan` and controller that represents and tracks the orphaned runtime instances. The controller deletes the runtime instance if receives a deletion request. The orphan CR reflects the existence of the runtime instance.

Reviewer: Do we introduce a new controller in the implementation?

Author: No, we're going to reuse existing orphan controller. Sorry for misleading and will refine it.

> 1. Introduce some new type of existing CRD `orphan` and controller that represents and tracks the orphaned runtime instances. The controller deletes the runtime instance if receives a deletion request. The orphan CR reflects the existence of the runtime instance.

Reviewer: ```suggestion
1. Introduce new types of existing CRD `orphan` and controller that represents and tracks the orphaned runtime instances. The controller deletes the runtime instance if receives a deletion request. The orphan CR reflects the existence of the runtime instance.
```

> 1. Introduce some new type of existing CRD `orphan` and controller that represents and tracks the orphaned runtime instances. The controller deletes the runtime instance if receives a deletion request. The orphan CR reflects the existence of the runtime instance.

Reviewer: > The controller deletes the runtime instance if receives a deletion request.

Which controller?

Author: The orphan controller will take the responsibility to ask instance manager to delete the instance process. Will reword this part.

    - Users can delete the orphaned runtime instances directories by `kubectl -n longhorn-system delete orphan <name>`.
    - Users can enable or disable the global auto-deletion by `kubectl -n longhorn-system edit settings orphan-engine-instance-auto-deletion`
    - Users can enable or disable the global auto-deletion by `kubectl -n longhorn-system edit settings orphan-replica-instance-auto-deletion`

## Design

### Implementation Overview

- Orphan CRs track the liveness of corresponding orphaned runtime instance
- Orphan CRs are created by instance manager monitor
- Orphan CRs are deleted by
  - Longhorn node controller: when node is deleted or evicted
  - Instance manager monitor:
    - When runtime instance disappear, delete the corresponding orphan runtime instance CR.
    - When auto deletion is enabled, delete all orphan runtime instance CRs on this node.
    - When runtime instance is rescheduled back to the node, delete the corresponding orphan runtime instance CR since it is no longer orphaned.

**Settings**

  - Add setting `orphan-engine-instance-auto-deletion`. Default value is `false`.

>   - Add setting `orphan-replica-instance-auto-deletion`. Default value is `false`.

Reviewer 2: NIT: Consider consolidating this into a single setting `orphan-instance-auto-deletion`, which accepts a list with possible values of `engine, replica`. If no values are provided, the behavior is no-op.

Author: > If no values are provided, the behavior is no-op.

Do you mean not perform auto deletion when the value is empty/invalid?

Reviewer 2: If we decided to go with the list approach, input should be restricted to valid values, limiting the acceptable inputs to `engine` and `replica`. When the list is empty, Longhorn should behave as same as when both `orphan-engine-instance-auto-deletion` and `orphan-replica-instance-auto-deletion` are set to `false`.

Reviewer: @c3y1huang Sorry, I'm a bit confused. What is the list approach?

Reviewer 2: I am thinking if we should consider using a single setting `orphan-instance-auto-deletion`, which takes a list as value, instead of having 2 separate settings `orphan-engine-instance-auto-deletion` and `orphan-replica-instance-auto-deletion` that take boolean as value. WDYT?

Reviewer: After discussing with @c3y1huang, we have a setting `orphan-auto-deletion` that takes boolean as a value and only handles the orphaned replica directories.

We can deprecate the `orphan-auto-deletion` setting and introduce a new setting, `orphan-resource-auto-deletion`, which supports values like `replica` (replica directory), `engine instance`, and `replica instance` (you can adjust the naming). This approach can avoid introducing additional settings in the future.

WDYT @COLDTURNIP

Author: A list here would be more flexible. To this approach, I suggest to use a comma-separate string, and accept the following three possible items:
- `replicaData` for replica data store
- `engineInstance` for engine runtime instance
- `replicaInstance` for replica runtime instance

Reviewer: > A list here would be more flexible. To this approach, I suggest to use a comma-separate string, and accept the following three possible items:
> 
> * `replicaData` for replica data store
> * `engineInstance` for engine runtime instance
> * `replicaInstance` for replica runtime instance

LGTM.

Reviewer 2: > To this approach, I suggest to use a comma-separate string

Other settings use a semicolon as the separator; consider doing the same for consistency.

  - Record the data engine type in `Parameters["dataEngine"]`.
  - Record the runtime instance's state in `status.conditions[].status` with type `"InstanceState"`

**Orphan controller**

  Reconciles the orphan events.

  - If the `deletionTimestamp` is non-zero, and the current controller is responsible for this orphan CR, proceed with deletion request.
    - If the orphan's node ID is the current controller ID:
      - Check with the corresponding engine or replica CR before deleting the runtime instance.
        - If the engine / replica CR not exists, this runtime instance is deletable.
        - If the engine / replica CR exists, and the node ID is not the current controller ID, then the runtime instance is deletable.
        - Once the runtime instance is deletable, create an instance manager client from the instance manager CR, and delete the runtime instance.
      - If the instance disappears from the instance manager, remove the finalizer to delete the orphan CR.
    - Otherwise, the orphan runtime instance lives on another node, but the controller loses its ownership because of disconnection. Remove the finalizer to complete deletion.

**Engine and replica controller**

>   - Once an engine or replica CR's node ID is assigned, delete the corresponding orphan runtime instance CR if exists on assigned node.

Reviewer: Can you elaborate more on the purpose?

Author: Sure. Once an engine or replica CR is scheduled to a node, and there's an existing instance process on this node for this engine/replica CR, then this instance is no longer considered to be orphaned. If there's an existing orphan CR for this instance, the orphan CR should be deleted without removing the instance process. This part should be judged in the instance monitor instead of engine/replica controller. Will refine in the LEP.

**Longhorn node controller**

  Reconcile the Longhorn node events.

  - Delete orphan CRs on a deleted or evicted node

**longhorn-ui**
    
  - Allow users to list the orphans on the node page by sending `OrphanList` call to the backend. 
  - Allow users to select the orphans to be deleted. The frontend needs to send `OrphanDelete` call to the backend.
    

### Test Plan

**Integration tests**

- `orphan` CRs will be created correctly while the process instance created unexpectedly. And they can be cleaned up without touching the data store.
- `orphan` CRs will be created correctly to a re-connected node. And they can be cleaned up without touching the data store.
- `orphan` CRs will be removed when an engine / replica is relocated back to the node.
- `orphan` CRs will be removed when the node is evicted or down.
- Auto-deletion setting.

> ## Note[optional]

Reviewer: Add Orphan CR LEP here. Readers can understand the context more.

    - Users can choose the items in the orphaned replica directory list and clean up them.
    - Users can enable the global auto-deletion on setting page. By default, the auto-deletion is disabled.

- Via `kubectl`
    - Users can list the orphaned runtime instances directories by `kubectl -n longhorn-system get orphans`.
    - Users can delete the orphaned runtime instances directories by `kubectl -n longhorn-system delete orphan <name>`.
    - Users can enable or disable the global auto-deletion by `kubectl -n longhorn-system edit settings orphan-engine-instance-auto-deletion`
    - Users can enable or disable the global auto-deletion by `kubectl -n longhorn-system edit settings orphan-replica-instance-auto-deletion`

## Design

### Implementation Overview

- Orphan objects track the liveness of corresponding orphaned runtime instance
- Orphan objects are created by instance manager monitor
- Orphan objects are deleted by
  - instance manager controller: when auto deletion is enabled
  - engine and replica controller: when runtime instance is rescheduled back to the node
  - Longhorn node controller: when node is deleted or evicted

>   - Instance manager monitor: when runtime instance disappear

Reviewer 2: I feel we should minimize the handling of orphaned engine/replica deletions across multiple controllers to avoid potential conflicts. Is it possible to limit this responsibility to the orphan controller only?

Author: Since the instance monitor has both the watching stream and a 60sec polling period, it is possible to merge the setting and engine/replica reschedule trigger into instance manager monitor:

1. Orphan instance judgement, then
2. Create new orphan objects, then
3. Delete existed or created objects because of
    1. auto deletion enabled
    2. engine or replica rescheduled back to this node
    3. runtime instance disappear

And keep the deletion of node evicted/deleted in node controller, just the same place we delete orphaned data store. WDYT?

**Settings**

  - Add setting `orphan-engine-intstance-auto-deletion`. Default value is `false`.
  - Add setting `orphan-replica-intstance-auto-deletion`. Default value is `false`.

**Instance manager controller**

  - Start the instance monitor during initialization.
  - Reconciles the settings events and orphan events.
    - When auto deletion enabled, delete the orphan runtime instance objects on this node.

**Instance manager monitor**

  - Establishes a GRPC stream with instance manager to watch the status of runtime instances.
  - Receives engine instance status update events from GRPC stream
  - Update runtime instance status on the list in instance manager object's `status.instanceEngines` and `status.instanceReplicas`
  - Compare the known engine/replica runtime instances with corresponding engine/replica objects:

>     - If there's no such object, mark this instance as orphaned.

Author: It is actually marked on Kubernetes object. In this step, we filter the orphaned instances. Will reword it:

> If there's no such object, this instance is considered to be orphaned.

    - If there's an object:
      - If the current state is not the desired state, ignore it.
      - If the object is in running state:
        - If the owner ID is not node ID, ignore it.
        - Mark the instance orphaned if and only if the node ID is not the current controller ID.
      - If the object is in stopped state:
        - Mark the instance orphaned if and only if the node ID is not the current controller ID.
      - To other object states, ignore it because of unstable state.

>   - Compare the orphaned engine/replica runtime instances with the exist orphan objects

Reviewer 2: Is there a possibility of false identification during orphan detection when the network is not stable?

Author: Confirmed that engine/replica instances will be recovered if it died or be deleted accidentally. So the problem is to prevent keeping deleting instances on orphan controller. Once the instance disappeared, the instance manager monitor should also update the instance state on corresponding orphan CR, hence the orphan controller is able to remove the finalizer directly.

Reviewer: > We try to delete the engine on node-1
> Before we finished delete the engine, engine.Spec.NodeID quickly move back to node-1

When the engine moves back to node-1 and before a new engine process is created, will the old engine process be deleted? Or will both the new and old engine processes coexist in the instance manager pod?

Author: > When the engine moves back to node-1 and before a new engine process is created, will the old engine process be deleted? Or will both the new and old engine processes coexist in the instance manager pod?

To our current v1.8.1 implementation, when a engine is scheduled back to the node, the old engine process will be killed first, then launch another engine process.

Reviewer: Because the race condition caused by engine deletion and engine creation occurs on the same node, I'm thinking if we can introduce a per-node read-write lock for 
- read lock: for engine creation 
- write lock: for engine deletion

During the write lock protection, controller can check whether the engine process is really deletable.

Author: It would be technically possible to implement a lock between engine/replica and orphan controllers. Let me clarify the detail.

* while initiating a deletion, engine or orphan controller acquire the write lock, and will release after instance manager confirmed deletion completed
* while initiating a creation, engine controller acquire the read lock, and will release after instance manager confirmed creation completed
* the lock and unlock events should be recorded
* if a controller failed to acquire the lock, raise an conflict error to re-enqueue the event again

    - Create a corresponding orphan object for each orphaned instance if missing.

>     - Delete the exist orphan runtime instance objects if the runtime instance disappeared from instance manager

Reviewer 2: I find the use of the word `object` in this doc a bit confusing - are you referring to Kubernetes Custom Resources (CRs), or a cached in-memory object used by the controllers? Consider using CR when referring to Kubernetes Custom Resources.

     │             ◄──────┤  desired node ? │     │                     │            
     └──────▲──────┘      │                 │     │                     │            
            │             └────────┬────────┘     │                     │            
            │                  yes │              │                     │            
            │                      │              │                     │            
            │              ┌───────▼──────────────▼───────┐  ┌──────────▼───────────┐
            │              │                              │  │                      │
            │              │ object node == current node? │  │ instance is orphaned │
            │              │                              │  │                      │
            │              └───────┬──────────────┬───────┘  └──────────▲───────────┘
            └──────────────────────┘              └─────────────────────┘            
               object on this node                  object on other node             
  ```

**Orphan object**

  - The object ID is calculated from engine/replica ID.
    - To engine instance: `orphan-engine-${checksum}`

>     - To replica instance: `orphan-replica-${checksum}`

Author: I'd like to know the reason of appending a random string. Wouldn't it more convenient to a determined name that can be calculated from the name of instance?

>     - To replica instance: `orphan-replica-${checksum}`

Reviewer: > Wouldn't it more convenient to a determined name that can be calculated from the name of instance?

The `random string` can be calculated from the name of the instance as well.

>     - To engine instance: `orphan-engine-${checksum}`
>     - To replica instance: `orphan-replica-${checksum}`

Author: How about 
```suggestion
    - To engine instance: `orphan-${engine-name}-${checksum}`
    - To replica instance: `orphan-${replica-name}-${checksum}`
```

    - `$checksum = sha256("${engine_replica_id}-${node_id}-${data_engine_type}")`
  - labels:
    - `longhorn.io/component`: `orphan`
    - `longhorn.io/managed-by`: `longhorn-manager`
    - `longhorn.io/orphan-type`: `engine-instance` or `replica-instance`
    - `longhornnode`: node ID
    - `longhornengine`: the instance name. Set only when it is an engine instance
    - `longhornreplica`: the instance name. Set only when it is a replica instance
  - The `OrphanType` is `engine-instance` or `replica-instance`.
  - Record the data engine type in `Parameters["dataEngine"]`.

**Orphan controller**

  Reconciles the orphan events.

  - If the `deletionTimestamp` is non-zero, and the current controller is responsible for this orphan object, proceed with deletion request.
    - If the orphan's node ID is the current controller ID:
      - Check with the corresponding engine or replica object before deleting the runtime instance.

>         - if the engine / replica object not exists, ask instance manager to delete the runtime instance directly.

Reviewer 2: > ask instance manager to delete the runtime instance directly

Could you clarify how this request is made or triggered?

Author: Orphan controller first fetch the instance manager CR, and try to create a client of instance manager (`engineapi.NewInstanceManagerClient`). With the information recorded in orphan CR, including the instance name and the data engine type, orphan controller can invoke `instanceManagerClient.InstanceDelete` with empty `diskUUID` and `cleanupRequired` disabled.

    - Users can enable the global auto-deletion on setting page. By default, the auto-deletion is disabled.

- Via `kubectl`
    - Users can list the orphaned runtime instances directories by `kubectl -n longhorn-system get orphans`.
    - Users can delete the orphaned runtime instances directories by `kubectl -n longhorn-system delete orphan <name>`.
    - Users can enable or disable the global auto-deletion by `kubectl -n longhorn-system edit settings orphan-engine-instance-auto-deletion`
    - Users can enable or disable the global auto-deletion by `kubectl -n longhorn-system edit settings orphan-replica-instance-auto-deletion`

## Design

### Implementation Overview

- Orphan CRs track the liveness of corresponding orphaned runtime instance
- Orphan CRs are created by instance manager monitor
- Orphan CRs are deleted by
  - Longhorn node controller: when node is deleted or evicted
  - Instance manager monitor:
    - When runtime instance disappear, delete the corresponding orphan runtime instance CR.
    - When auto deletion is enabled, delete all orphan runtime instance CRs on this node.

>     - When runtime instance is rescheduled back to the node, delete the corresponding orphan runtime instance CR since it is no longer orphaned.

Reviewer 3: The presence of an orphan engine instance can block node draining, which in turn causes the Harvester upgrade process to get stuck in [Phase 4](https://docs.harvesterhci.io/v1.5/upgrade/troubleshooting#phase-4-upgrade-nodes). We've [encountered this issue multiple times](https://docs.harvesterhci.io/v1.5/upgrade/v1-4-1-to-v1-4-2#1-upgrade-is-stuck-in-the-pre-drained-state).

With this enhancement, will the orphan engine be automatically cleaned up as part of the node draining process?

Reviewer 3: No, Harvester does not perform node eviction. Instead, it relies on Longhorn volume live migration.

> When an engine is rescheduled and left the node, instance monitor must be able to identify & delete the orphaned engine process in this design.

The IM handles this in real time? For example, if node draining is blocked by an orphan engine, how long does it usually take for the Longhorn manager to detect and remove the orphan engine?

Also, Harvester performs node draining sequentially—one node at a time during the upgrade process.

Author: Once the volume is attaching to the new node, the IM handles process deletion as soon as the engine controller receives the old engine deletion, and the IM monitor (mentioned in this LEP) will also receives the process status changing event immediately from the GRPC event stream. Also, IM monitor has a process status polling period every minute, hence the process exiting will be detected withing 60sec in worst case. In this LEP, the orphan judgement occurs each time when the IM monitor receives and polls the status of node processes.
